libraries/apollo-ksp/src/main/kotlin/com/apollographql/apollo3/ksp/ksp-validation.kt

```diff
@@ -0,0 +1,165 @@
+import com.apollographql.apollo3.compiler.ir.IrOptionalType
+import com.apollographql.apollo3.compiler.ir.IrScalarType
+import com.apollographql.apollo3.compiler.ir.IrType
+import com.apollographql.apollo3.compiler.resolveSchemaType
+import com.google.devtools.ksp.symbol.KSTypeReference
+import com.google.devtools.ksp.symbol.Nullability
+
+class IncompatibleType(message: String) : Exception(message)
+
+@OptIn(ApolloInternal::class)
+internal class ValidationScope(
+    private val objectMapping: Map<String, ObjectInfo>,
+    private val scalarMapping: Map<String, ScalarInfo>,
+    private val schema: Schema,
+    private val codegenMetadata: CodegenMetadata
+) {
+  fun validateAndCoerce(ksTypeReference: KSTypeReference, expectedType: GQLType, allowCovariant: Boolean): IrType {
+    val ksType = ksTypeReference.resolve()
+    val className = ksType.declaration.acClassName()
+
+    var expectedNullableType = expectedType
+    if (expectedType is GQLNonNullType) {
+      expectedNullableType = expectedType.type
+    }
+
+    val irType = when (expectedNullableType) {
+      is GQLListType -> {
+        if (className != listClassName) {
+          throw IncompatibleType("Expected list type at ${ksTypeReference.location}")
+        }
+        IrListType(
+            validateAndCoerce(
+                ksTypeReference.element!!.typeArguments.single().type!!,
+                expectedNullableType.type,
+                allowCovariant,
+            )
+        )
+      }
+
+      is GQLNamedType -> {
+        when (val typeDefinition = schema.typeDefinition(expectedNullableType.name)) {
+          is GQLScalarTypeDefinition -> {
+            val scalarInfo = scalarMapping.get(typeDefinition.name)
+            if (scalarInfo == null) {
+              throw IncompatibleType("Expected scalar type '${typeDefinition.name}' but no adapter found. Did you forget a @ApolloAdapter? at ${ksTypeReference.location}")
+            }
+            if (scalarInfo.targetName != className.asString()) {
+              throw IncompatibleType("Scalar type '${typeDefinition.name}' is mapped to '${scalarInfo.targetName} but '${className.asString()} was found at ${ksTypeReference.location}")
+            }
+            IrScalarType(typeDefinition.name)
+          }
+
+          is GQLInputObjectTypeDefinition -> {
+            val expectedFQDN = codegenMetadata.resolveSchemaType(typeDefinition.name)?.toIrClassName() ?: error("Cannot resolve input ${typeDefinition.name}")
+            if (className != expectedFQDN) {
+              throw IncompatibleType("Input object type '${typeDefinition.name}' is mapped to '${expectedFQDN} but '${className.asString()} was found at ${ksTypeReference.location}")
+            }
+
+            IrInputObjectType(typeDefinition.name)
+          }
+
+          is GQLEnumTypeDefinition -> {
+            val expectedFQDN = codegenMetadata.resolveSchemaType(typeDefinition.name)?.toIrClassName() ?: error("Cannot resolve enum ${typeDefinition.name}")
```

Comment:
This is the actual change. We used to rely on `CodegenLayout` but this can be much better modeled using `CodegenMedata` instead.